core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,481 @@
+            data[program_id]['k_gram_hashes'], window_size)
+    return data
+
+
+def _calc_jaccard_index(A, B):
+    """Calculate jaccard's coefficient for two sets A and B."""
+    small_set = A[:] if len(A) < len(B) else B[:]
+    union_set = B[:] if len(A) < len(B) else A[:]
+    for elem in union_set:
+        if elem in small_set:
+            small_set.remove(elem)
+    union_set.extend(small_set)
+
+    if union_set == []:
+        return 0
+
+    small_set = A[:] if len(A) < len(B) else B[:]
+    intersection_set = []
+    for elem in small_set:
+        if elem in A and elem in B:
+            intersection_set.append(elem)
+            A.remove(elem)
+            B.remove(elem)
+
+    coeff = float(len(intersection_set)) / len(union_set)
+    return coeff
+
+
+def _get_program_similarity(fingerprint_a, fingerprint_b):
+    """Find similarity between fingerprint of two programs."""
+    A = [h for (h, _) in fingerprint_a]
+    B = [h for (h, _) in fingerprint_b]
+    return _calc_jaccard_index(A, B)
+
+
+def _generate_top_similars(data, top):
+    """Find 'top' nearest neighbours for all programs in dataset."""
+    for program_id_1 in data:
+        overlaps = []
+        for program_id_2 in data:
+            overlap = _get_program_similarity(
+                data[program_id_1]['fingerprint'],
+                data[program_id_2]['fingerprint'])
+            overlaps.append((program_id_2, overlap))
+        data[program_id_1]['top_similar'] = sorted(
+            overlaps, key=lambda e: e[1], reverse=True)[:top]
+    return data
+
+
+def _run_knn(data):
+    """Predict classes for each program in dataset using KNN."""
+
+    # No. of times a class has to appear in nearest neighbours of a program
+    # so that prediction is correct.
+    occurrence = 0
+
+    # Keep record of missclassified programs' ID.
+    missclassified_points = []
+
+    for pid in data:
+        similars = data[pid]['top_similar']
+        if pid in similars:
+            similars.remove(pid)
+        nearest_classes = [data[i]['class'] for (i, _) in similars]
+        cnt = Counter(nearest_classes)
+        common = cnt.most_common(1)
+        data[pid]['prediction'] = common[0][0]
+
+        if data[pid]['prediction'] == data[pid]['class']:
+            # If prediction is correct then add no. of times the correct
+            # class has to appear in nearest neighbours to occurrence.
+            occurrence += common[0][1]
+        else:
+            # Else mark the ID fo program so that it will be
+            # trained using SVM later.
+            missclassified_points.append(pid)
+
+    # Calculate average of occurrence. This is used during prediction so
+    # that if winner of nearest neighbours appears less than the occurrence
+    # we consider that KNN has failed in prediction.
+    occurrence /= float(len(data))
+
+    return occurrence, missclassified_points
+
+
+# pylint: disable=too-many-instance-attributes, attribute-defined-outside-init
```

Comment:
Pylint puts limit of 7 instance attributes and I need 8 instance attributes so that later on we can invoke `to_dict()` method to get `classifier_data`. Also these attributes are declared during training because we can't know their value before training so we can't declare them during `init`.